File: android/java/org/chromium/chrome/browser/crypto_wallet/AssetRatioServiceFactory.java
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.brave_wallet.mojom.AssetRatioService;
import org.chromium.chrome.browser.crypto_wallet.util.Utils;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
public class AssetRatioServiceFactory {
    private static final Object lock = new Object();
    private static AssetRatioServiceFactory instance;

    public static AssetRatioServiceFactory getInstance() {
        synchronized (lock) {
            if (instance == null) {
                instance = new AssetRatioServiceFactory();
            }
        }
        return instance;
    }

    private AssetRatioServiceFactory() {}

    public AssetRatioService getAssetRatioService(ConnectionErrorHandler connectionErrorHandler) {
        Profile profile = Utils.getProfile(false); // always use regular profile
        long nativeHandle =
                AssetRatioServiceFactoryJni.get().getInterfaceToAssetRatioService(profile);
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        AssetRatioService assetRatioService = AssetRatioService.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) assetRatioService).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return assetRatioService;
    }

    private MessagePipeHandle wrapNativeHandle(long nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        long getInterfaceToAssetRatioService(Profile profile);
    }
}

File: ios/browser/brave_wallet/asset_ratio_service_factory.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

#include "brave/ios/browser/brave_wallet/asset_ratio_service_factory.h"

#include "base/no_destructor.h"
#include "brave/components/brave_wallet/browser/asset_ratio_service.h"
#include "components/keyed_service/core/keyed_service.h"
#include "ios/chrome/browser/shared/model/profile/profile_ios.h"
#include "ios/chrome/browser/shared/model/profile/profile_manager_ios.h"
#include "ios/web/public/browser_state.h"
#include "services/network/public/cpp/shared_url_loader_factory.h"

namespace brave_wallet {

// static
mojo::PendingRemote<mojom::AssetRatioService>
AssetRatioServiceFactory::GetForProfile(ProfileIOS* profile) {
  return GetInstance()
      ->GetServiceForProfileAs<AssetRatioService>(profile, true)
      ->MakeRemote();
}

// static
AssetRatioService* AssetRatioServiceFactory::GetServiceForState(
    ProfileIOS* profile) {
  return GetInstance()->GetServiceForProfileAs<AssetRatioService>(profile,
                                                                  true);
}

// static
AssetRatioServiceFactory* AssetRatioServiceFactory::GetInstance() {
  static base::NoDestructor<AssetRatioServiceFactory> instance;
  return instance.get();
}

AssetRatioServiceFactory::AssetRatioServiceFactory()
    : ProfileKeyedServiceFactoryIOS("AssetRatioService",
                                    ProfileSelection::kRedirectedInIncognito,
                                    ServiceCreation::kCreateLazily,
                                    TestingCreation::kCreateService) {}

AssetRatioServiceFactory::~AssetRatioServiceFactory() = default;

std::unique_ptr<KeyedService> AssetRatioServiceFactory::BuildServiceInstanceFor(
    ProfileIOS* profile) const {
  std::unique_ptr<AssetRatioService> asset_ratio_service(
      new AssetRatioService(profile->GetSharedURLLoaderFactory()));
  return asset_ratio_service;
}

}  // namespace brave_wallet
